feat(iam): SCRUM-51 Create User on Registration

adnlv opened the pull request:
**Jira**: [SCRUM-51](https://epam-team-php.atlassian.net/browse/SCRUM-51)

## Description

Introduce a user creation endpoint in the application's IAM.

### Fixes

- Allow `php-cs-fixer` risky rules in `cs:check` composer command to pass CI
- Rename `phpstan` container XML path to match current project name
- Extend `phpstan` memory limit to pass the CI
- Apply migrations on PHP container startup

## How to roll back?

- Rollback migrations:

  ```shell
  docker compose exec php php bin/console doctrine:migrations:execute --no-interaction --down "Twitter\\IAM\\Infrastructure\\Persistence\\MySQL\\Migrations\\Version20251217062759"
  ```

  <img width="1200" height="153" alt="Screenshot_20251217_083653" src="https://github.com/user-attachments/assets/74bf6ff9-d449-4be3-9bb4-97b25e0f3c4f" />

- Revert this pull request

## How has this been tested?

Unit tests cover domain and application layers.

To reproduce, run:

```shell
docker compose exec php ./vendor/bin/phpunit tests/IAM --display-all-issues --testdox --group unit
```

You should see the following output with all tests passed:

```
PHPUnit 12.5.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.4.16
Configuration: /app/phpunit.dist.xml

......................... 25 / 36 ( 69%)
...........               36 / 36 (100%)

Time: 00:07.152, Memory: 14.00 MB

Create User Command Handler (Twitter\Tests\IAM\Application\CreateUser\CreateUserCommandHandler)
 ✔ Creates a user and returns an identifier
 ✔ Fails when email is invalid
 ✔ Fails when password is invalid
 ✔ Fails when user already exists
 ✔ Propagates unexpected repository failures

Email (Twitter\Tests\IAM\Domain\User\Model\Email)
 ✔ Can be created from a valid email with data set "simple email"
 ✔ Can be created from a valid email with data set "with subdomain"
 ✔ Can be created from a valid email with data set "with plus alias"
 ✔ Can be created from a valid email with data set "numeric local part"
 ✔ Can be created from a valid email with data set "dash and dot"
 ✔ Throws when created from an invalid email string
 ✔ Two emails with the same value are equal
 ✔ Emails with different values are not equal

Password Hash (Twitter\Tests\IAM\Domain\User\Model\PasswordHash)
 ✔ Hashes and verifies a valid plain password
 ✔ Rejects a too short plain password
 ✔ Creates a value object from existing hash
 ✔ Rejects an invalid password hash
 ✔ Validates plain password format with data set "valid password"
 ✔ Validates plain password format with data set "short password"
 ✔ Validates plain password format with data set "no uppercase letter"
 ✔ Validates plain password format with data set "no lowercase letter"
 ✔ Validates plain password format with data set "no number"
 ✔ Validates plain password format with data set "no special character"
 ✔ Validates plain password format with data set "valid with a different special char"
 ✔ Validates plain password format with data set "valid with a dash and underscore"
 ✔ Verifies plain passwords correctly with data set "valid password"
 ✔ Verifies plain passwords correctly with data set "empty password"
 ✔ Verifies plain passwords correctly with data set "short password"
 ✔ Verifies plain passwords correctly with data set "password with no uppercase letter"
 ✔ Verifies plain passwords correctly with data set "password with no lowercase letter"
 ✔ Verifies plain passwords correctly with data set "password with no number"

User Id (Twitter\Tests\IAM\Domain\User\Model\UserId)
 ✔ Generates a non empty identifier
 ✔ Can be recreated from its string representation
 ✔ Throws when created from an invalid string
 ✔ Two ids with the same value are equal
 ✔ Two generated ids are not equal

OK (36 tests, 42 assertions)
```

### Manual testing

#### Create user successfully

Request

```json
{
    "email": "test@example.com",
    "password": "qwerty123"
}
```

Response

```json
{
    "id": "019b35db-ce74-767e-bba4-374f835e4622"
}
```

#### Failed attempt to create user with invalid email

Request

```json
{
    "email": "test@example",
    "password": "qwerty123"
}
```

Response

```json
{
    "error": {
        "code": "INVALID_EMAIL",
        "message": "The provided email address is invalid"
    }
}
```

#### Failed attempt to create user with invalid password

Request

```json
{
    "email": "test@example.com",
    "password": "qwerty1"
}
```

Response

```json
{
    "error": {
        "code": "INVALID_PASSWORD",
        "message": "The provided password does not meet requirements"
    }
}
```

#### Failed attempt to create a user which already exists
Request

```json
{
    "email": "test@example.com",
    "password": "qwerty123"
}
```

Response

```json
{
    "error": {
        "code": "USER_ALREADY_EXISTS",
        "message": "A user with this email already exists"
    }
}
```

## Media attachments

adnlv committed:
chore(dependencies): SCRUM-51 Update dependencies to latest versions

Require `symfony/uid` package for UUID generation.

adnlv committed:
fix(phpstan): SCRUM-51 Update container XML path for renamed kernel

adnlv committed:
style(cs-fixer): SCRUM-51 Normalize formatting in PHP-CS-Fixer config

adnlv committed:
feat(iam): SCRUM-51 Add user domain models

User value objects handle business rules validation.

adnlv committed:
feat(shared): SCRUM-51 Add custom DBAL types for user VOs

There are required to map user value objects to persistence data.

adnlv committed:
feat(iam): SCRUM-51 Implement MySQL user repository

adnlv committed:
feat(iam): SCRUM-51 Add XML mapping for user entity

adnlv committed:
feat(iam): SCRUM-51 Add MySQL migration for users table

adnlv committed:
feat(iam): SCRUM-51 Add user creation API endpoint

adnlv committed:
feat(iam): SCRUM-51 Add exceptions for email and password validation

adnlv committed:
refactor(iam): SCRUM-51 Remove unused imports in exception classes

adnlv committed:
refactor(iam): SCRUM-51 Replace `findByEmail` with `existsByEmail`

The new name better represents what method is used for within the use
case.

adnlv committed:
feat(iam): SCRUM-51 Add roles and enforce unique email constraint

Roles added to match the documented database schema. User email address
must be unique due to business requirements.

adnlv committed:
refactor(iam): SCRUM-51 Improve `InvalidEmailException` error message

Error message now includes the invalid email for better debugging.

adnlv committed:
refactor(iam): SCRUM-51 Add command-based user registration flow

adnlv committed:
fix(phpstan): SCRUM-51 Extend memory limit to pass CI

adnlv committed:
fix(iam): SCRUM-51 Handle unexpected errors in user creation command
handler

adnlv committed:
chore(phpunit): SCRUM-51 Ignore PHPUnit cache files

adnlv committed:
test(iam): SCRUM-51 Add unit tests for `UserId` validation and generation

adnlv committed:
test(iam): SCRUM-51 Add unit tests for `Email` validation

adnlv committed:
test(iam): SCRUM-51 Add unit tests for `PasswordHash` validation

adnlv committed:
test(iam): SCRUM-51 Add unit tests for `CreateUserCommandHandler`

adnlv committed:
refactor(iam): SCRUM-51 Handle password hashing exceptions

adnlv committed:
chore(doctrine): SCRUM-51 Set default charset and collate

adnlv committed:
refactor(tests): SCRUM-51 Update namespaces for IAM-related test files

adnlv committed:
chore(php-cs-fixer): SCRUM-51 Update cache file path to match PHPUnit config

adnlv committed:
chore(autoload): SCRUM-51 Remove redundant IAM namespace

adnlv committed:
refactor(iam): SCRUM-51 Rename `save` to `add` in `UserRepository`

"add" explicitly indicates that a new entity is being inserted into
persistence, while "save" typically handles both inserting and
updating entities.

adnlv committed:
refactor(iam): SCRUM-51 Rename properties in `Email` and `UserId`

adnlv committed:
refactor(iam): SCRUM-51 Replace custom Doctrine types with native mappings

Despite providing a configurable way to map value object into
persistence data, custom types add complexity on top of features
implementation and maintenance. Try to use native fields for scalar
values and embeddable for value objects with multiple fields.

adnlv committed:
refactor(iam): SCRUM-51 Use binary ID for users table to optimize storage

adnlv committed:
chore(core): SCRUM-51 Clean up redundant configuration comments

adnlv committed:
chore(doctrine): SCRUM-51 Remove deprecated Doctrine entity mapping dir

adnlv committed:
test(iam): SCRUM-51 Add `#[CoversClass]` for test coverage mapping

adnlv committed:
feat(iam): SCRUM-51 Add centralized exception handling

adnlv committed:
fix(docker): SCRUM-51 Allow migration execution when no changes detected

adnlv committed:
refactor(iam): SCRUM-51 Update consistent HTTP exception handling

adnlv committed:
test(iam): SCRUM-51 Add `tearDown` method to clean up test state

adnlv committed:
feat(iam): SCRUM-51 Integrate Monolog for logging and error tracking

Logs are necessary to track occurred errors, because they may have
sensitive information, that a client do not permitted to see. Therefore,
choose Monolog as it is the most popular(*) PHP logging library, which
Symfony integrates seamlessly with. It fully covers both the project's
requirements and future use cases.

(*) According to the official Symfony documentation:
https://symfony.com/doc/current/logging.html#monolog

adnlv committed:
refactor(iam): SCRUM-51 Simplify `InvalidPasswordException` constructor

For now any field except `message` is useless.

adnlv committed:
style(php-cs-fixer): SCRUM-51 Follow PER `braces_position` standard

According to the PER Coding Style 3.0(*):
"If a function or method contains no statements or comments
(such as an empty no-op implementation or when using constructor
property promotion), then the body SHOULD be abbreviated as `{}` and
placed on the same line as the previous symbol, separated by a space"

(*) https://www.php-fig.org/per/coding-style/#44-methods-and-functions

adnlv committed:
refactor(iam): SCRUM-51 Mark `CreateUserCommandHandler` as readonly

Command handlers are supposed to be immutable.

adnlv committed:
fix(php-cs-fixer): SCRUM-51 Replace nonexistent rules

adnlv committed:
style(php-cs-fixer): SCRUM-51 Simplify FQCNs with imported classes

adnlv committed:
refactor(iam): SCRUM-51 Remove redundant `toString` methods

adnlv committed:
refactor(iam): SCRUM-51 Remove `get` prefix from getter methods

adnlv committed:
refactor(iam): SCRUM-51 Use heredoc for SQL in migrations

adnlv committed:
refactor(iam): SCRUM-51 Optimize queries and update migrations

adnlv committed:
style(php-cs-fixer): SCRUM-51 Apply method chaining indentation rule

adnlv committed:
refactor(iam): SCRUM-51 Remove array wrapping for single event listener

adnlv committed:
refactor(iam): SCRUM-51 Simplify error response creation

adnlv committed:
fix(php-cs-fixer): SCRUM-51 Allow risky rules to pass CI

adnlv committed:
fix(php-cs-fixer): SCRUM-51 Allow risky rules on `check`, not `fix`

adnlv committed:
refactor(iam): SCRUM-51 Use count query in `existsByEmail`

`SELECT` seems a bit more optimized, but as a trade-off for
*unmeasured* performance improvement, it throws a bunch of exceptions,
that the application has to deal with. `count()` on the other hand,
throws no exceptions and fits perfectly for `exists*` methods.

adnlv committed:
feat(iam): SCRUM-51 Register `ExceptionSubscriber` service

adnlv commented:
This PR continues the [previous](https://github.com/CPAprince/Twitter/pull/17) conversation. See the [last comment](https://github.com/CPAprince/Twitter/pull/17#issuecomment-3678841381) for an explanation.

CPAprince commented on src/IAM/Domain/User/Model/PasswordHash.php:
```diff
@@ -0,0 +1,54 @@
+
+    public static function fromHash(string $hash): self
+    {
+        if (strlen($hash) < 60) {
```
`if (strlen($hash) === 60 && strpos($hash,"$2y$") === 0)`

better check for PASSWORD_BCRYPT

CPAprince commented on src/IAM/Domain/User/Model/PasswordHash.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    public static function fromPlainPassword(string $plainPassword): self
+    {
+        if (strlen($plainPassword) < 8) {
```
`if(preg_match("/^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*\-_\.]).{8,}$/m", $plainPassword))`

better check min length and min 1 number, 1 upcase letter and 1 low case letter

https://regex101.com/r/oZ6r5G/1

CPAprince reviewed (approved):
Pay attention to suggested checks in comments

adnlv committed:
build(composer): SCRUM-51 Update dependencies

adnlv committed:
feat(iam): SCRUM-51 Make password hash validation more strict

adnlv committed:
feat(iam): SCRUM-51 Add more strict password requirements

adnlv committed:
fix(iam): SCRUM-51 Update unit tests to match validation requirements

adnlv committed:
refactor(iam): SCRUM-51 Make immutable properties readonly

adnlv committed:
build(iam): SCRUM-51 Require Symfony's serializer bundle

It is required to map request to objects and inject them into invokable
controllers.

adnlv committed:
feat(iam): SCRUM-51 Add lazy validation for user creation requests

adnlv committed:
refactor(iam): SCRUM-51 Removed redundant checks and catches

User uniqueness should be enforced at the database level, catching and
handling `UniqueConstraintViolationException`.

adnlv committed:
refactor(tests): SCRUM-51 Rename methods and use valid password constant

adnlv committed:
test(iam): SCRUM-51 Simplify UserId tests and improve coverage

adnlv committed:
test(iam): SCRUM-51 Add Email tests cases and improve assertions

adnlv committed:
refactor(tests): SCRUM-51 Remove "it" prefix from test method names

adnlv committed:
test(iam): SCRUM-51 Rename user creation command tests

adnlv committed:
refactor(iam): SCRUM-51 Remove unused `UserAlreadyExistsException`

adnlv committed:
refactor(iam): SCRUM-51 Map lazy assertion errors in dedicated methods

adnlv committed:
feat(iam): SCRUM-51 Check password hash format

CPAprince reviewed (approved):
Seems to be OK

ihpr commented on composer.json:
```diff
@@ -7,15 +7,23 @@
         "doctrine/doctrine-migrations-bundle": "^3.7",
-        "doctrine/orm": "^3.5.8",
+        "doctrine/orm": "^3.6.0",
+        "phpdocumentor/reflection-docblock": "^5.6.6",
```
@adnlv Why do we need this package?

ihpr commented on src/IAM/Domain/User/Model/Email.php:
```diff
@@ -0,0 +1,32 @@
+
+namespace Twitter\IAM\Domain\User\Model;
+
+use Twitter\IAM\Domain\User\Model\Exception\InvalidEmailException;
```
@adnlv (low) Exception is not a part of the domain model as it has no business behaviour.
```suggestion
use Twitter\IAM\Domain\User\Exception\InvalidEmailException;
```

ihpr commented on src/IAM/Infrastructure/Persistence/MySQL/Migrations/Version20251217102810.php:
```diff
@@ -0,0 +1,39 @@
+
+declare(strict_types=1);
+
+namespace Twitter\IAM\Infrastructure\Persistence\MySQL\Migrations;
```
@adnlv (low) Most of the folder names are in singular not in plural.
```suggestion
namespace Twitter\IAM\Infrastructure\Persistence\MySQL\Migration;
```

ihpr commented on src/IAM/Domain/User/Model/PasswordHash.php:
```diff
@@ -0,0 +1,55 @@
+     */
+    public static function fromPlainPassword(string $plainPassword): self
+    {
+        // https://regex101.com/r/oZ6r5G/1
```
@adnlv (low) What if the link will expire? Better to cover it with a good unit test.

ihpr commented on src/IAM/Domain/User/Model/UserId.php:
```diff
@@ -0,0 +1,37 @@
+        return new self($id);
+    }
+
+    public function __toString(): string
```
@adnlv (low) Magic methods goes before other publics methods.

ihpr commented on src/IAM/Infrastructure/Persistence/MySQL/Migrations/Version20251217102810.php:
```diff
@@ -0,0 +1,39 @@
+            CREATE TABLE users (
+                roles JSON NOT NULL,
+                email VARCHAR(320) NOT NULL,
+                password_hash VARCHAR(60) NOT NULL,
```
```suggestion
                password_hash CHAR(60) NOT NULL,
```
@adnlv (medium) Please also check the mapping, so the diff won't change the structure.

ihpr commented on config/routes.yaml:
```diff
@@ -1,12 +1,10 @@
+api_users_create:
+    path: /api/users
+    controller: Twitter\IAM\UI\REST\Controller\CreateUserController
+    methods: POST
```
@adnlv (low) Attributes type is ok to use here.

ihpr commented on src/IAM/Application/CreateUser/CreateUserCommandHandler.php:
```diff
@@ -0,0 +1,35 @@
+     * @throws InvalidEmailException
+     * @throws InvalidPasswordException
+     */
+    public function handle(CreateUserCommand $command): CreateUserCommandResult
```
@adnlv (low) That's how the DTO is used. `CreateUserCommandResult` is the DTO. It transfer the model data from Application layer to Infrastructure (or Presentation) without exposing entity methods or sensitive info (like HashedPassword).

ihpr commented on src/IAM/Domain/User/Model/PasswordHash.php:
```diff
@@ -0,0 +1,55 @@
+        try {
+            $hash = password_hash($plainPassword, PASSWORD_BCRYPT);
+        } catch (Error $error) {
+            throw new RuntimeException('Unable to hash password: '.$error->getMessage(), previous: $error);
```
@adnlv Why RuntimeException? Why not InvalidPasswordException with a different message?

adnlv replied:
@ihpr I think that using `InvalidPasswordException` instead would be incorrect, as a failure in `password_hash()` indicates a runtime or environment error, rather than an invalid password provided by the user that violates domain rules.

ihpr commented on src/IAM/Domain/User/Model/PasswordHash.php:
```diff
@@ -0,0 +1,55 @@
+
+    public static function fromHash(string $hash): self
+    {
+        if (60 !== strlen($hash) || !str_starts_with($hash, '$2y$')) {
```
```suggestion
        if (self::HASH_LENGTH !== strlen($hash) || !str_starts_with($hash, '$2y$')) {
```
(low)

ihpr commented on src/IAM/Infrastructure/Persistence/MySQL/Repository/MySQLUserRepository.php:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    public function existsByEmail(string $email): bool
```
@adnlv Where do you sue this method? I don't see you call it. Maybe I'm blind.

ihpr commented on src/IAM/Domain/User/Model/UserRepository.php:
```diff
@@ -0,0 +1,12 @@
+{
+    public function add(User $user): void;
+
+    public function existsByEmail(string $email): bool;
```
@adnlv (low) Can be better with `Email` value object.
```suggestion
    public function existsByEmail(Email $email): bool;
```

ihpr commented on tests/IAM/Domain/User/Model/EmailTest.php:
```diff
@@ -0,0 +1,55 @@
+class EmailTest extends TestCase
+{
+    #[Test]
+    public function canBeCreatedFromAValidEmailString(): void
```
```suggestion
    public function canBeCreatedFromAValidEmail(): void
```
@adnlv (medium) To improve this test I would suggest creating a [Data Provider](https://docs.phpunit.de/en/12.5/attributes.html#dataprovider) with few correct variants.

ihpr commented on tests/IAM/Domain/User/Model/EmailTest.php:
```diff
@@ -0,0 +1,55 @@
+        $a = Email::fromString($value);
+        $b = Email::fromString($value);
+
+        self::assertEquals((string) $a, (string) $b);
```
```suggestion
        self::assertSame((string) $a, (string) $b);
```

ihpr commented on tests/IAM/Domain/User/Model/EmailTest.php:
```diff
@@ -0,0 +1,55 @@
+        $a = Email::fromString('a@example.com');
+        $b = Email::fromString('b@example.com');
+
+        self::assertNotEquals((string) $a, (string) $b);
```
```suggestion
        self::assertNotSame((string) $a, (string) $b);
```

ihpr commented on tests/IAM/Domain/User/Model/PasswordHashTest.php:
```diff
@@ -0,0 +1,84 @@
+#[CoversClass(PasswordHash::class)]
+class PasswordHashTest extends TestCase
+{
+    private const string VALID_PASSWORD = 'Pswd.123';
```
@adnlv (medium) DRY principle doesn't work with tests as it affects isolation. Use [DUMP](https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html). It means this value is ok to duplciate from test to test. It makes better isolation and explicit the values used for tesitng.

ihpr commented on tests/IAM/Domain/User/Model/PasswordHashTest.php:
```diff
@@ -0,0 +1,84 @@
+    {
+        self::expectException(InvalidArgumentException::class);
+
+        PasswordHash::fromHash('abracadabra');
```
lol

ihpr commented on tests/IAM/Domain/User/Model/PasswordHashTest.php:
```diff
@@ -0,0 +1,84 @@
+        self::assertSame($expected, $passwordHash->verify($input));
+    }
+
+    public static function passwordVerificationProvider(): array
```
@adnlv There is a performance optimization is to use `Generator` as the return type for Data Provider.

Same thing with invalid values.

```suggestion
    public static function passwordVerificationProvider(): Generator
```

ihpr commented on tests/IAM/Domain/User/Model/PasswordHashTest.php:
```diff
@@ -0,0 +1,84 @@
+    public static function passwordVerificationProvider(): array
+    {
+        return [
+            'valid password' => [self::VALID_PASSWORD, true],
```
```suggestion
            yield 'valid password' => [self::VALID_PASSWORD, true];
```

ihpr commented on tests/IAM/Domain/User/Model/UserIdTest.php:
```diff
@@ -0,0 +1,62 @@
+    #[Test]
+    public function canBeRecreatedFromItsStringRepresentation(): void
+    {
+        $original = UserId::generate();
```
@adnlv (medium) Set the value explicitly to improve test isolation. `fromString` tests shouldn't depend on how the `generate` method works.

ihpr reviewed (approved):
@adnlv Good work I see a great improvement.

Tests looks awesome. There are jsut few corrections I addressed to make it ideal.

I also foudn that you declared the method in repository but didn't use it. That's my biggest concern.

Please check. Afetr addressing those, no more reviews from my end is required.

adnlv committed:
refactor(iam): SCRUM-51 Move domain exceptions out of `Model` namespace

Exceptions have no business logic and therefore they are not part of
domain models.

adnlv committed:
refactor(iam): SCRUM-51 Use singular directory name instead of plural

adnlv committed:
refactor(iam): SCRUM-51 Place magic methods before other public ones

adnlv committed:
refactor(iam): SCRUM-51 Follow DUMP principle in unit tests

It makes better isolation and clearly shows the values used for testing.

adnlv committed:
refactor(composer): SCRUM-51 Remove unused `phpdocumentor` dependencies

adnlv committed:
refactor(iam): SCRUM-51 Remove redundant `existsByEmail` method

Database became the end source of truth, therefore this method serves
no purpose.

adnlv committed:
refactor(iam): SCRUM-51 Define a password hash length constant

adnlv committed:
refactor(iam): SCRUM-51 Use stricter asserts for comparisons

adnlv committed:
refactor(iam): SCRUM-51 Use attributes for API routes

adnlv committed:
refactor(iam): SCRUM-51 Enforce fixed-length password hash

adnlv committed:
refactor(iam): SCRUM-51 Use generators for test data providers

adnlv committed:
refactor(iam): SCRUM-51 Add data provider for email tests

adnlv committed:
test(iam): SCRUM-51 Cover plain password validation with unit tests

adnlv committed:
refactor(iam): SCRUM-51 Remove outdated regex reference comment

adnlv committed:
refactor(shared): SCRUM-51 Move event subscribers to Shared module
